efibootmanager_policy.go

```diff
@@ -0,0 +1,335 @@
+func AddEFIBootManagerProfile(profile *PCRProtectionProfile, params *EFIBootManagerProfileParams) error {
+	// Load event log
+	eventLog, err := os.Open(eventLogPath)
+	if err != nil {
+		return xerrors.Errorf("cannot open TCG event log: %w", err)
+	}
+	log, err := tcglog.NewLog(eventLog, tcglog.LogOptions{})
+	if err != nil {
+		return xerrors.Errorf("cannot parse TCG event log header: %w", err)
+	}
+
+	if !log.Algorithms.Contains(tcglog.AlgorithmId(params.PCRAlgorithm)) {
+		return errors.New("cannot compute secure boot policy digests: the TCG event log does not have the requested algorithm")
+	}
+
+	profile.AddPCRValue(params.PCRAlgorithm, bootManagerCodePCR, make(tpm2.Digest, params.PCRAlgorithm.Size()))
+
+	// Replay the event log until we see the transition from "pre-OS" to "OS-present". The event log may contain measurements
+	// for system preparation applications, and spec-compliant firmware should measure a EV_EFI_ACTION “Calling EFI Application
+	// from Boot Option” event before the EV_SEPARATOR event, but not all firmware does this.
+	for {
+		event, err := log.NextEvent()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return xerrors.Errorf("cannot parse TCG event log: %w", err)
+		}
+
+		if event.PCRIndex != bootManagerCodePCR {
+			continue
+		}
+
+		profile.ExtendPCR(params.PCRAlgorithm, bootManagerCodePCR, tpm2.Digest(event.Digests[tcglog.AlgorithmId(params.PCRAlgorithm)]))
+		if event.EventType == tcglog.EventTypeSeparator {
+			break
+		}
+	}
+
+	root := bootManagerCodePolicyGenBranch{profile: profile}
+	allBranches := []*bootManagerCodePolicyGenBranch{&root}
+
+	var loadEvents []*bmLoadEventAndBranch
+	var nextLoadEvents []*bmLoadEventAndBranch
+
+	if len(params.LoadSequences) == 1 {
+		loadEvents = append(loadEvents, &bmLoadEventAndBranch{event: params.LoadSequences[0], branch: &root})
+	} else {
+		for _, e := range params.LoadSequences {
+			branch := root.branch()
+			allBranches = append(allBranches, branch)
+			loadEvents = append(loadEvents, &bmLoadEventAndBranch{event: e, branch: branch})
+		}
+	}
+
+	for len(loadEvents) > 0 {
+		e := loadEvents[0]
+		loadEvents = loadEvents[1:]
+
+		digest, err := computePeImageDigest(params.PCRAlgorithm, e.event.Image)
+		if err != nil {
+			return err
+		}
+		e.branch.profile.ExtendPCR(params.PCRAlgorithm, bootManagerCodePCR, digest)
+
+		if len(e.event.Next) == 1 {
+			e.event = e.event.Next[0]
+			nextLoadEvents = append(nextLoadEvents, e)
+		} else {
+			for _, n := range e.event.Next {
+				branch := e.branch.branch()
+				nextLoadEvents = append(nextLoadEvents, &bmLoadEventAndBranch{event: n, branch: branch})
+				allBranches = append(allBranches, branch)
+			}
+		}
+
+		if len(loadEvents) == 0 {
+			loadEvents = nextLoadEvents
+			nextLoadEvents = nil
+		}
+	}
+
+	for _, b := range allBranches {
+		if len(b.branches) == 0 {
+			continue
+		}
+		b.profile.AddProfileOR(b.branches...)
+	}
```

Comment:
It happens on the first iteration of this loop, where b == root.

Reply:
ah, I somehow read {&root} much earlier as {}.  Anyway after that I stared at the loop some more, maybe it merits a comment that the order of processing allBranches doesn't matter because we are building a tree of OR instructions, not computing something.